core/src/com/unciv/models/ruleset/tile/TileResource.kt

```diff
@@ -1,29 +1,125 @@
 import com.unciv.models.translations.tr
 import java.util.*
 
+/**
+ * Class representing a Resource.
+ * 
+ * This is deserialized from json, therefore is has a default constructor
+ */
 class TileResource : NamedStats() {
 
     var resourceType: ResourceType = ResourceType.Bonus
     var terrainsCanBeFoundOn: List<String> = listOf()
     var improvement: String? = null
     var improvementStats: Stats? = null
     var revealedBy: String? = null
+
+    @Deprecated("As of 3.15.x", ReplaceWith("uniques"))
     var unique: String? = null
+    var uniques = ArrayList<String>()
+    val uniqueObjects:List<Unique> by lazy {
+        // Compatibility code for deprecated [unique]
+        if (unique != null) {
+            if (unique!!.isNotEmpty()) uniques.add(unique!!)
+            unique = null
+        }
+        uniques.map { Unique(it) } 
+    }
+
+    /** Can map generator/editor add this resource to a given [tile]?
+     * @param tile The [TileInfo] to check
+     * @return `true` = resource is allowed there, `false` = don't place there
+     */
+    fun isAllowedOnTile(tile: TileInfo): Boolean {
+        // uniques / terrainsCanBeFoundOn: Cannot unique should have highest precedence,
+        // while terrainsCanBeFoundOn and Can unique should be treated like a set union
+
+        getMatchingUniques("Cannot occur on [] tiles").forEach { 
+            if (tile.matchesUniqueFilter(it.params[0])) return false
+        }
+        getMatchingUniques("Can occur on [] tiles").forEach {
+            if (tile.matchesUniqueFilter(it.params[0])) return true
+        }
+
+        // From old code...
+        if (tile.baseTerrain == Constants.snow && tile.isHill()) return false  // todo: this should not be hardcoded
+        // Original logic:
+        return isAllowedOnTerrain(tile.getLastTerrain())
+    }
+
+    /** Should [TileInfo.normalizeToRuleset] remove this resource from a given [tile]?
+     * @param tile The tile to check
+     * @return `true` if normalization should remove the resource 
+     */
+    fun mustRemoveFromTile(tile: TileInfo): Boolean {
+        if (isAllowedOnTerrain(tile.baseTerrain)) return false
+        return tile.terrainFeatures.none { isAllowedOnTerrain(it) }
+    }
+
+    /** Is this resource allowed on a given [terrain]? */
+    private fun isAllowedOnTerrain(terrain: Terrain): Boolean {
+        return isAllowedOnTerrain(terrain.name)
+    }
 
+    /** Is this resource allowed on a [terrain] given as String? */
+    fun isAllowedOnTerrain(terrain: String): Boolean {
+        // todo: Called from Terrain.getDescription - how to support tileFilter=terrain uniques?
+        return terrain in terrainsCanBeFoundOn
+    }
+
+    /** Get a terrain the map editor can use as background
+     * @param ruleset The ruleset to search for viable terrains
+     * @return Terrain name or null if impossible due to [ruleset] inconsistency 
+     */
+    fun getSampleTerrain(ruleset: Ruleset): String? {
```

Comment:
For one usage? I don't think this is needed.

Reply:
I know. That's got to do with "general structure" - I generalized as much as possible first, planning to fold such stuff later if rarely (or not at all) used...